discovery/ecs/client/aws.go

```diff
@@ -0,0 +1,541 @@
+			return nil, err
+		}
+
+		for _, ci := range resp2.ContainerInstances {
+			contInsts = append(contInsts, ci)
+		}
+
+		if resp.NextToken == nil || aws.StringValue(resp.NextToken) == "" {
+			break
+		}
+		params.NextToken = resp.NextToken
+
+	}
+	log.Debugf("Retrieved %d container instances on cluster %s", len(contInsts), aws.StringValue(cluster.ClusterName))
+	return contInsts, nil
+}
+
+func (c *AWSRetriever) getTasks(cluster *ecs.Cluster) ([]*ecs.Task, error) {
+	tasks := []*ecs.Task{}
+
+	params := &ecs.ListTasksInput{
+		Cluster:    cluster.ClusterArn,
+		MaxResults: aws.Int64(maxAPIRes),
+	}
+	log.Debugf("Getting cluster %s tasks", aws.StringValue(cluster.ClusterName))
+
+	// Start listing tasks
+	for {
+		resp, err := c.ecsCli.ListTasks(params)
+		if err != nil {
+			return nil, err
+		}
+
+		tArns := []*string{}
+		for _, t := range resp.TaskArns {
+			tArns = append(tArns, t)
+		}
+
+		// Describe tasks
+		params2 := &ecs.DescribeTasksInput{
+			Cluster: cluster.ClusterArn,
+			Tasks:   tArns,
+		}
+
+		resp2, err := c.ecsCli.DescribeTasks(params2)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, t := range resp2.Tasks {
+			// We don't want stopped tasks
+			if t.StoppedAt == nil {
+				tasks = append(tasks, t)
+			}
+		}
+
+		if resp.NextToken == nil || aws.StringValue(resp.NextToken) == "" {
+			break
+		}
+		params.NextToken = resp.NextToken
+
+	}
+	log.Debugf("Retrieved %d tasks on cluster %s", len(tasks), aws.StringValue(cluster.ClusterName))
+	return tasks, nil
+}
+
+func (c *AWSRetriever) getServices(cluster *ecs.Cluster) ([]*ecs.Service, error) {
+	srvs := []*ecs.Service{}
+
+	params := &ecs.ListServicesInput{
+		Cluster:    cluster.ClusterArn,
+		MaxResults: aws.Int64(maxAPIRes),
+	}
+	log.Debugf("Getting cluster %s services", aws.StringValue(cluster.ClusterName))
+	// Start listing services
+	for {
+		resp, err := c.ecsCli.ListServices(params)
+		if err != nil {
+			return nil, err
+		}
+
+		srvArns := []*string{}
+		for _, srv := range resp.ServiceArns {
+			srvArns = append(srvArns, srv)
+		}
+
+		// Describe services
+		params2 := &ecs.DescribeServicesInput{
+			Cluster:  cluster.ClusterArn,
+			Services: srvArns,
+		}
+
+		resp2, err := c.ecsCli.DescribeServices(params2)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, srv := range resp2.Services {
+			srvs = append(srvs, srv)
+		}
+
+		if resp.NextToken == nil || aws.StringValue(resp.NextToken) == "" {
+			break
+		}
+		params.NextToken = resp.NextToken
+
+	}
+	log.Debugf("Retrieved %d services on cluster %s", len(srvs), aws.StringValue(cluster.ClusterName))
+	return srvs, nil
+}
+
+func (c *AWSRetriever) getTaskDefinitions(tDIDs []*string, useChache bool) ([]*ecs.TaskDefinition, error) {
+	tDefs := []*ecs.TaskDefinition{}
+	var mErr error
+	var wg sync.WaitGroup
+	cached := 0
+	for _, tID := range tDIDs {
+		// If we want cache then check if already present
+		if _, ok := c.cache.getTaskDefinition(aws.StringValue(tID)); useChache && ok {
```

Comment:
This call happens during the concurrent goroutines building the cache. The cache getters are not protected by a mutex though so this is a race.